activerecord/test/cases/transaction_callbacks_test.rb

```diff
@@ -449,6 +449,42 @@ def test_before_commit_update_in_same_transaction
   end
 end
 
+class CallbacksOnIdempotentActionsTest < ActiveRecord::TestCase
+  class TopicWithHistory < ActiveRecord::Base
+    self.table_name = :topics
+
+    def self.history
+      @@history ||= []
+    end
+  end
+
+  class TopicWithIdempotentCallbacksOnDestroy < TopicWithHistory
+    after_commit(on: :destroy, idempotent: true) { |record| record.class.history << :destroy }
+  end
+
+  class TopicWithNonIdempotentCallbacksOnDestroy < TopicWithHistory
+    after_commit(on: :destroy, idempotent: false) { |record| record.class.history << :destroy }
+  end
+
+  def test_non_idempotent_callback_on_destroy
+    test_helper(TopicWithNonIdempotentCallbacksOnDestroy)
+    assert_equal [:destroy], TopicWithNonIdempotentCallbacksOnDestroy.history
+  end
+
+  def test_idempotent_callback_on_destroy
+    test_helper(TopicWithIdempotentCallbacksOnDestroy)
+    assert_equal [:destroy, :destroy], TopicWithIdempotentCallbacksOnDestroy.history
+  end
+
+  def test_helper(klass)
```

Comment:
I would name that differently. Something like: `simulate_race_condition` or something like that.